design/src/docs/anatomy.md

```diff
@@ -0,0 +1,601 @@
+    router: RoutingService<RestRouter<S>, AwsRestJson1>,
+}
+```
+
+## Plugins
+
+Smithy Rust also provides a way to hook into the upgrade procedure in order to modify the service behavior. This is done via the `Plugin` trait:
+
+```rust
+/// A mapping from one [`Operation`] to another. Used to modify the behavior of
+/// [`Upgradable`](crate::operation::Upgradable) and therefore the resulting service builder,
+///
+/// The generics `Protocol` and `Op` allow the behavior to be parameterized.
+///
+/// Every service builder enjoys [`Pluggable`] and therefore can be provided with a [`Plugin`] using
+/// [`Pluggable::apply`].
+pub trait Plugin<Protocol, Op, S, L> {
+    type Service;
+    type Layer;
+
+    /// Maps an [`Operation`] to another.
+    fn map(&self, input: Operation<S, L>) -> Operation<Self::Service, Self::Layer>;
+}
+```
+
+The `Upgradable::upgrade` method on `Operation<S, L>`, previously presented in [Upgrading a Model Service](#upgrading-a-model-service), is more accurately:
+
+```rust
+    /// Takes the [`Operation<S, L>`](Operation), applies [`Plugin`], then applies [`UpgradeLayer`] to
+    /// the modified `S`, then finally applies the modified `L`.
+    ///
+    /// The composition is made explicit in the method constraints and return type.
+    fn upgrade(self, plugin: &Pl) -> Self::Service {
+        let Operation { inner, layer } = plugin.map(self);
+        let layer = Stack::new(UpgradeLayer::new(), layer);
+        layer.layer(inner)
+    }
+```
+
+![Upgradable Diagram](imgs/upgradable-plugin.png)
+
+An example `Plugin` implementation can be found in [aws-smithy-http-server/examples/pokemon-service/src/plugin.rs](https://github.com/awslabs/smithy-rs/blob/main/rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/plugin.rs).
+
+All builders implement the `Pluggable` trait, which allows them to apply plugins to service builders:
+
+```rust
+/// Provides a standard interface for applying [`Plugin`]s to a service builder. This is implemented automatically for
+/// all builders.
+///
+/// As [`Plugin`]s modify the way in which [`Operation`]s are [`upgraded`](crate::operation::Upgradable) we can use
+/// [`Pluggable`] as a foundation to write extension traits which are implemented for all service builders.
+///
+/// # Example
+///
+/// ```
+/// # struct PrintPlugin;
+/// # use aws_smithy_http_server::plugin::Pluggable;
+/// trait PrintExt: Pluggable<PrintPlugin> {
+///     fn print(self) -> Self::Output where Self: Sized {
+///         self.apply(PrintPlugin)
+///     }
+/// }
+///
+/// impl<Builder> PrintExt for Builder where Builder: Pluggable<PrintPlugin> {}
+/// ```
+pub trait Pluggable<NewPlugin> {
+    type Output;
+
+    /// Applies a plugin to the service builder.
+    fn apply(self, plugin: NewPlugin) -> Self::Output;
+}
+```
+
+As seen in the `Pluggable` documentation, third-parties can use extension traits over `Pluggable` to extend the API of builders.
+
+## Accessing Unmodelled Data
```

Comment:
This section is quite brief. The full story amounts to pepper the code with `Exts: FromParts<P>` in a predictable way and can be read quite easily from the source code.

Should I go into more detail? Should I include the paragraph above?